### PR TITLE
make usage of prebuilt go binaries opt-in only

### PR DESCRIPTION
**Pull request by EvanBoyle**
This change forces the specification of a `runtime.options.prebuilt` property in order for the go language host to use a prebuilt executable.

Fixes https://github.com/pulumi/pulumi/issues/4123

```yaml
name: ls
runtime:
    name: go
    options: 
        prebuilt: myprebuilt
description: A minimal Go Pulumi program
```
```sh
$pulumi up
Previewing update (dev):
     Type                 Name    Plan     Info
     pulumi:pulumi:Stack  ls-dev           1 error
 
Diagnostics:
  pulumi:pulumi:Stack (ls-dev):
    error: expected to find prebuilt executable: unable to find program: myprebuilt
```

```sh

$go build -o myprebuilt
$pulumi up 
Previewing update (dev):
     Type                 Name    Plan       
 +   pulumi:pulumi:Stack  ls-dev  create     
 
Resources:
    + 1 to create

Do you want to perform this update? yes
Updating (dev):
     Type                 Name    Status      
 +   pulumi:pulumi:Stack  ls-dev  created     
 
Resources:
    + 1 created

Duration: 1s

Permalink: https://app.pulumi.com/EvanBoyle/ls/dev/updates/1
```

@mikhailshilkov can you provide input here for consistency with dotnet? I figure that this property could be used in dotnet in the future to execute a prebuilt dll via `dotnet myprebuilt.dll`. Does that sound right to you? I understand that this doesn't solve for specification of a csproj file to build from, but that can be a separate property.

### CODE REVIEW

**Comment by EvanBoyle** [comment]
added documentation here: https://github.com/pulumi/docs/pull/2892

**Comment by mikhailshilkov** [comment]
One thing I'm questioning a bit is the name `prebuilt`. Maybe something with `executable` or `binary`?

**Comment by lukehoban** [comment]
> One thing I'm questioning a bit is the name `prebuilt`. Maybe something with `executable` or `binary`?

FWIW - almost noted the same.  My order of preference is probably this (though don't feel super-strongly):
1. `binary`
2. `executable`
3. `prebuilt`

**Comment by EvanBoyle** [comment]
@lukehoban @mikhailshilkov dotnet DLLs are not executables from what I understand. Would you still prefer `binary` or `executable`?

**Comment by EvanBoyle** [comment]
@lukehoban Where is the test coverage that you're talking about? We use this in an example, but it doesn't look like it really anything.

File: sdk/go/common/resource/plugin/langruntime_plugin.go [hunk 1]
```diff
@@ -57,7 +57,12 @@ func NewLanguageRuntime(host Host, ctx *Context, runtime string,
 
 	var args []string
 	for k, v := range options {
-		args = append(args, fmt.Sprintf("-%s=%t", k, v))
+		fmtstr := "-%s=%t"
```
Comment (pgavlin) [inline sdk/go/common/resource/plugin/langruntime_plugin.go:60]:
```suggestion
		fmtstr := "-%s=%v"
```

File: sdk/go/common/resource/plugin/langruntime_plugin.go [hunk 2]
```diff
@@ -57,7 +57,12 @@ func NewLanguageRuntime(host Host, ctx *Context, runtime string,
...
+		// the prebuilt option specifies a string with an executable name
+		if k == "prebuilt" {
+			fmtstr = "-%s=%s"
+		}
```
[the remark this answers is not in the document]
Comment (lukehoban) [reply sdk/go/common/resource/plugin/langruntime_plugin.go:64]: Happy to go with `$v` instead per @pgavlin suggestion.

**Comment by lukehoban** [comment]
> Where is the test coverage that you're talking about? We use this in an example, but it doesn't look like it really anything.

Yeah - looks like we may not actually have been using it correctly in the one place we have a test that includes it.  We may need to add test harness support for `binary: ` and `typescript: false` to get validation of this - but happy to do that in a follow-up.

**Pull request merged**